sympy/tensor/array/nditer.py

```diff
@@ -0,0 +1,53 @@
+from sympy.tensor.array.dense_ndim_array import DenseNDimArray
+from sympy.tensor.array.sparse_ndim_array import SparseNDimArray
+from sympy.tensor.array.ndim_array import NDimArray
+from sympy.core.compatibility import Iterable
+
+class nditer(object):
+
+    def __init__(self, iterable):
+        from sympy.matrices.matrices import MatrixBase
+
+        if not isinstance(iterable, (Iterable, MatrixBase)):
+            raise NotImplementedError("Data type not yet supported")
+        self._iter = iterable
+        self._idx = 0
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        from sympy.matrices.matrices import MatrixBase
+
+        try:
+            if isinstance(self._iter, list):
+                result = self._iter[self._idx]
```

Comment:
How to dynamically handle the iteration over a multi-dimension list is a great challenge. Especially for the case where an `array` is cast to `list`:
```python
>>> a = Array([[1, 2], [3, 4]])
>>> b = list(a)
>>> b.__class__
<class 'list'>
>>> b[0].__class__
<class 'sympy.tensor.array.dense_ndim_array.ImmutableDenseNDimArray'>
```
I might need some guidance in this problem. :-)

Reply:
Please ignore `list`. We have `Array` in SymPy. People who use `list` will use 99% of the time `NumPy`, we don't need support for `list`.

Reply:
We don't need to redo what NumPy already does... if you want to use NumPy's features, just use NumPy.

Reply:
But we have tests like:
```
 assert list(sparse_array) == [0, 0, 0, 1]
```
That is why I ran into an error about `list` and tried to handle this special case. However, I have changed the test by using a simple `tolist()` instead of 'list()'.